### DISCUSSION TITLE
How to get raw sql for migrations

### DISCUSSION THREAD

**Discussion by Alksar** [fluentmigrator/fluentmigrator#1625] [Q&A] [answered]
Hi, thanks for the project.

How to properly get raw sql for migrations in fluent style?

**Comment by ademchenko** [comment]
Hello, @Alksar , your question is unclear. Please, give an example of what you want.

**Comment by jzabroski** [comment, marked as answer]
Hi @Alksar ,

If you want to insert custom raw sql into an expression, you can use `Raw.Sql(string)` signature.  This is used for escaping data within insert, update and delete expressions.

If you want to get the raw sql generated for a migration, then we call that a _script_.  There is a command line option to generate a script via Preview Mode.

If you want something else, please explain like @ademchenko suggested.  I'm going to move this Issue to a Discussion since it's not really an issue.

**Reply by GoodSimonVM** [reply]
Hi! Let's imagine how I wrote some fluent migration. Now I want to get a raw SQL script to be executed by FluentMigrator. How can I get or view this script?

**Reply by jzabroski** [reply]
This belongs in a separate question. But you can use a connectionless migration to get the raw SQL script unless your migration relies on Execute.WithConnection and therefore has a feedback loop where you are able to inspect the contents of the database prior to building the script. Otherwise, you would need to customize how FluentMigrator announces/prints what is being executed.

Another option (that is a workaround), would be to use a profiling tool and capture the executed SQL. For example SQL Server Profiler will log this data to the TextData column for a few specific events that ship with the default template for SQL Server Profiler.  You can set an ApplicationName for your connection string and filter on that unique ApplicationName to avoid unwanted traffic from other applications.

**Reply by GoodSimonVM** [reply]
Could you describe the first method more precisely? With example, please.

**Reply by jzabroski** [reply]
@GoodSimonVM This is covered in the documentation.

**Reply by rklec** [reply]
@jzabroski Same idea here, I want to test in an assertion that my migration produces [actually the same SQL](https://stackoverflow.com/a/77947724/15488899) as some other tooling.

I did not found that in the documentation. Could you point me to a solution?

**Reply by jzabroski** [reply]
@rklec How will you account for subtle nuances like whitespace, etc?

If you want to know how to unit test Fluentmigrator, then did you consider reading a few of the thousands of tests? Tongue in cheek, but serious suggestion.

Other than that, I would just create two databases and run all your migrations against one, and your other tool against the other, and use a Schema Compare tool like DevArt for the last bits. I understand this isn't automated, but depends on your exact needs. I find it hard to imagine you want to continuously regress both tools, as that's twice the maintenance cost.

**Reply by rklec** [reply]
Well... whitespace ignoring is easy.

> Other than that, I would just create two databases and run all your migrations against one, and your other tool against the other, and use a Schema Compare tool like DevArt for the last bits. I understand this isn't automated, but depends on your exact needs. 

Yeah, actually it could be automated. I would also be able to use the raw SQL to create the DB schema and run integration tests against it. Or use FluentMigrator as usual, of course.
And I guess this is the idea, I'll be following. As a real integration test running against a real DB may be more useful, actually.

BTW [here is me arguing and fiddling with ChatGPT to ask it how to get the RAW SQL](https://chat.openai.com/share/06357ad8-f25b-4bb9-a009-275f35ae3312) for further reference.

**Reply by jzabroski** [reply]
It is possible ChatGPT is confused by Umbraco CMS code it may be aware of in its ll, which may have taken the Apache 2.0 license we offer to its extreme and created a non attributed fork.

Mildly cool, but Fluent NHibernate and other hallucinations suggest it would need fine tuning to be more powerful. Hopefully in ten years I don't get github issues from people complaining about how stupid Copilot is and they don't actually want to read code. Or maybe open source will twist and become trendy to support due to people not knowing otherwise.
